### PR TITLE
Feat: Create Auth0 OAuth Adapter

### PR DESCRIPTION
**Pull request by adityaoberai**
## What does this PR do?

Adds Auth0 OAuth Adapter

## Test Plan

- Enable Auth0 in the Users Settings tab
- Manually try logging in with Auth0

## Related PRs and Issues

None

### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?

Yes

### CODE REVIEW

**Comment by adityaoberai** [comment]
Here are some snippets from the Users page

image

Screenshot 2022-04-01 170143_censored

File: src/Appwrite/Auth/OAuth2/Auth0.php [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+    protected function getClientSecret(): string
+    {
+        $secret = $this->decodeJson();
+
+        return (isset($secret['clientSecret'])) ? $secret['clientSecret'] : ''; 
+    }
+
+     /**
+     * Extracts the Auth0 Domain from the JSON stored in appSecret. Defaults to 'common' as a fallback
+     * @return string
```
Comment (christyjacob4) [inline src/Appwrite/Auth/OAuth2/Auth0.php:186]: Have a new line here

File: src/Appwrite/Auth/OAuth2/Auth0.php [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+            $user = $this->request('GET', 'https://'.$this->getAuth0Domain().'/userinfo', $headers);
+            $this->user = \json_decode($user, true);
+        }
+
+        return $this->user;
+    }
+
+    /**
+     * Extracts the Client Secret from the JSON stored in appSecret
+     * @return string
```
Comment (christyjacob4) [inline src/Appwrite/Auth/OAuth2/Auth0.php:175]: Have a new line here

File: src/Appwrite/Auth/OAuth2/Auth0.php [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+        return $this->tokens;
+    }
+    
+    
+    /**
+     * @param string $refreshToken
+     *
+     * @return array
+     */
+    public function refreshTokens(string $refreshToken):array
```
Comment (christyjacob4) [inline src/Appwrite/Auth/OAuth2/Auth0.php:87]:
```suggestion
    public function refreshTokens(string $refreshToken): array
```
PSR-12 style guides
Screenshot 2022-04-02 at 4 18 50 PM

File: src/Appwrite/Auth/OAuth2/Auth0.php [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        
+        return '';
+    }
+    
+     /**
+     * @param string $accessToken
+     *
+     * @return array
+     */
+    protected function getUser(string $accessToken)
```
Comment (christyjacob4) [inline src/Appwrite/Auth/OAuth2/Auth0.php:162]:
```suggestion
    protected function getUser(string $accessToken): array
```

File: src/Appwrite/Auth/OAuth2/Auth0.php [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+    {
+        $secret = $this->decodeJson();
+        return (isset($secret['auth0Domain'])) ? $secret['auth0Domain'] : ''; 
+    }
+
+    /**
+     * Decode the JSON stored in appSecret
+     * @return array
+     */
+    protected function decodeJson(): array
```
Comment (christyjacob4) [inline src/Appwrite/Auth/OAuth2/Auth0.php:198]:
```suggestion
    protected function getAppSecret(): array
```

File: src/Appwrite/Auth/OAuth2/Auth0.php [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+    {    
+        try {
+            $secret = \json_decode($this->appSecret, true);
```
Comment (christyjacob4) [inline src/Appwrite/Auth/OAuth2/Auth0.php:201]:
```suggestion
            $secret = \json_decode($this->appSecret, true, 512, JSON_THROW_ON_ERROR);
```

File: src/Appwrite/Auth/OAuth2/Auth0.php [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+        } catch (\Throwable $th) {
+            throw new Exception('Invalid secret');
```
Comment (christyjacob4) [inline src/Appwrite/Auth/OAuth2/Auth0.php:203]:
```suggestion
            throw new \Exception('Invalid secret');
```

File: src/Appwrite/Auth/OAuth2/Auth0.php (hunk 6 above)
Comment (christyjacob4) [reply src/Appwrite/Auth/OAuth2/Auth0.php:201]: We need this flag, else json_decode wont throw any error.

File: src/Appwrite/Auth/OAuth2/Auth0.php [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+     */
+    protected function getAuth0Domain(): string
+    {
+        $secret = $this->decodeJson();
+        return (isset($secret['auth0Domain'])) ? $secret['auth0Domain'] : ''; 
+    }
+
+    /**
+     * Decode the JSON stored in appSecret
+     * @return array
```
Comment (christyjacob4) [inline src/Appwrite/Auth/OAuth2/Auth0.php:196]: Add a new b/w the description and the return doc

**Review by christyjacob4 — CHANGES_REQUESTED** [review]
Looks good 👍 
Some small changes.

**Comment by adityaoberai** [comment]
@christyjacob4 I have added the suggested changes and manually tested the OAuth adapter.

**Comment by adityaoberai** [comment]
> Awesome work 👏

Thank you 😁

**Comment by christyjacob4** [comment]
Waiting for tests....

**Comment by adityaoberai** [comment]
I did some manual QA earlier for the adapter.
Got a valid response in the created session.

```
{
	"$id": "6267d83aa26050478d98",
	"userId": "6267d83a983fe7f791f8",
	"expire": 1682508730,
	"provider": "auth0",
	"providerUid": "auth0|6267d8324cac35006c77c20f",
	"providerAccessToken": "eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIiwiaXNzIjoiaHR0cHM6Ly9hcHB3cml0ZS10ZXN0LmpwLmF1dGgwLmNvbS8ifQ..DdfrLMBpIzxzRXmh.Xb8zlPnqzTGgiwmklessHhdE54tFtopLge6kA7Yj5EUKJpHljr7_f4hcLFtLtSG3kYGIc7JOtxwt5vDvkgEUAqAMrdQVCw9wDdTFP2eXRzAt4-Kydm4zC-AsYVn8BAdAuBNM45oBqNNjNWn1yt-5GgHKq2JgRxzzL5wtYdzftlihI4gJpUIjZdZWOHmn31P4nJOEdKgal-TpJeoYwDwv7A1hXZ4V8BkkWemkb9EdgG7qqnlWAiNYL8PRU9vFqNhMbnbQK4U4xPlkWOQj7k5u8ItSapl27Rtj7vN7xkiMtBc7gsw7V0LJq2q62CxVod-c3H56Fj7aVapwsx-REqjPo9eU7w.toAVB6xL4tRrjuwSzyWqEw",
	"providerAccessTokenExpiry": 1651059130,
	"providerRefreshToken": "v1.MZwxdaibdfbmvO_JqOMomMA1Gw34XoT9FF-AFTIylSak_rP4ogLjUF0RT_32qF7VDkLJ1QVuZ0j8bI_f5gfOwJ4",
	"ip": "172.21.0.1",
	"osCode": "WIN",
	"osName": "Windows",
	"osVersion": "10",
	"clientType": "browser",
	"clientCode": "CH",
	"clientName": "Chrome",
	"clientVersion": "100.0",
	"clientEngine": "Blink",
	"clientEngineVersion": "",
	"deviceName": "desktop",
	"deviceBrand": "",
	"deviceModel": "",
	"countryCode": "--",
	"countryName": "Unknown",
	"current": true
}
```

**Review by christyjacob4 — APPROVED** [review]

**Pull request merged**